[Ong Wee, Marcus] iP

lfrostbytee opened the pull request:
# Biggest Heading
- Bullet List Point 1
- Bullet List Point 2
1. Number List Point A
2. Number List Point B
``` java
public void main() {
    console.log("haha");
}
```
- [x] task list point A 😀 
- [ ] task list point B 😇 
- [x] task list point C 🙃 
> cats can meow - by another cat
meow? [meow](https://www.google.com/search?q=cat&rlz=1C1CHBD_enSG938SG945&oq=cat&aqs=chrome..69i57j46i67i433j35i39j0i67l3j69i60l2.755j1j7&sourceid=chrome&ie=UTF-8)

`main` method is for cats

**bold**
*italics*
~~strikethrough~~

Commit:
Add Gradle support

lfrostbytee committed:
Level-1 Added

lfrostbytee committed:
Renamed Duke.java to Poolsheen.java

lfrostbytee committed:
Rewrote code to follow Java Style Guide

lfrostbytee committed:
Added "add and list" feature

lfrostbytee committed:
Changed variable names and re-formatted "list" command

lfrostbytee committed:
Submission for Level-3.
* Added mark command.
* Added unmark command.
* Created a new class called "Task".
* "list" now displays the status of each task.

lfrostbytee committed:
Add Deadline, Event, and ToDo

Tasks can be added through any non-keyword message

We want to create more categories for tasks to add more details
into our list.

Let's create a parent class called Task and create 3 other
subclasses: ToDo, Deadline and, Event.

Using inheritance is preferable because every task shares some
similar properties like a IsDone status and the description of the
task.

lfrostbytee committed:
Add automated testing

There is currently no automated testing in-place.

We want to ensure the Poolsheen program runs as expected without
needing to manually test every time a change is made.

Let's,
* create an input and expected file
* create a runtest.bat file which can be executed to run the tests

The INPUT.txt will be represent the inputs into the program.
The EXPECTED.TXT will be what the program should print and will be
compared with ACTUAL.TXT, the actual lines that our program will
print.

lfrostbytee committed:
Add error handling

There is currently no handling of edge cases when wrong
or incomplete inputs are entered.

We want Poolsheen to prompt the user on how they can fix their
wrong or incomplete inputs if an error occurs.

Let's,
*create an IncompleteCommandException class for incomplete commands
*create an UnknownCommandException class for unknown commands
*catch runtime errors with IndexOutOfBoundsException
and NumberFormatException
*catch unexpected errors using Exception

The IncompleteCommandException when caught will let the user know
what part of the command they are missing. The
UnknownCommandException lets the user know that the command does not
currently exist.

lfrostbytee committed:
Add delete task

There is currently no way to remove tasks that a user has
accidentally added into their Poolsheen program.

We want Poolsheen to include a "delete" command to delete a task
of position x from the list of tasks that Poolsheen remembers.

Let's create a new method called deleteTask which removes the task
of position x.

The position x will be fed after the input has been parsed. If
"delete 5" is inputted and the list of tasks has a size greater
than 5, the task of position 5 will be removed.

lfrostbytee committed:
Add enum types for commands

Commands are now only distinguished by the string commands that
they are assigned to.

We want each command to have a type using the Java enum type.

Let's create a new enum type for each command type.

The switch cases for processing the input will now compare using
the value of the string which will be converted to an enum type if
that enum type exists.

lfrostbytee committed:
Add saving and loading

Poolsheen currently does not remember tasks from when it was
previously run.

Poolsheen should remember tasks from previous runs because a user
may need to log off from Poolsheen to do other tasks.

Let's create a SAVE.TXT file which retains all information from
whenever Poolsheen runs and refer to it every time Poolsheen runs.

A .TXT file is static which allows information to be saved even
when Poolsheen stops running.

Constant names changed to follow Java Coding Standard.
.gitignore and runtest.bat modified to ignore SAVE.TXT files.

lfrostbytee committed:
Add understanding of dates and times.

Currently, random strings can be inputted for the dates and times
of deadlines and events.

Users may use Poolsheen ineffectively by inputting the wrong format
of date and time for deadlines and events.

The LocalDate class is introduced to replace the type of time from
String to LocalDate.

LocalDate offers several methods which allow us to compare dates
properly. Relying on LocalDate allows us to parse dates and times
properly as well.

lfrostbytee committed:
Merge branch 'branch-Level-7'

lfrostbytee committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	text-ui-test/EXPECTED.TXT

lfrostbytee committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

lfrostbytee committed:
Add Gradle

lfrostbytee committed:
Add more OOP

Currently most of the code resides in the Poolsheen.java file.

As the Poolsheen program expands, it becomes harder to maintain if
everything is kept inside one file. So, we can split what Poolsheen
can do into 4 classes.

Let's,
- create a Ui, Storage, Parser, and, TaskList class
- create an abstract Command class and its subclasses

We can now test the Poolsheen functionality based on each class
rather than the whole class. It is easier to change each command
as well.

lfrostbytee committed:
Add packaging

Currently, all the code resides in one directory

The code is difficult to maintain because it is difficult to
navigate and manage classes.

Let's,
- create a poolsheen package to encapsulate all classes
- create a command package to encapsulate all command classes
- create a task package to encapsulate all task classes

We can now separate Poolsheen's main functionalities from the
respective commands and types of tasks.

lfrostbytee committed:
Add JUnit tests

Currently, there is only automated UI text testing.

More forms of automated testing need to be included to ensure the
maintainability of the program.

Let's create unit tests for most classes of the Poolsheen program.

Unit tests allow developers of the Poolsheen program to isolate
bugs to their respective classes, making it easier to debug code.

lfrostbytee committed:
Package the Poolsheen App as a JAR file

The Poolsheen program can only be run on IntelliJ Idea by using the
in-built run button.

Poolsheen should also be able to be used by other users who do not
have IntelliJ Idea.

Let's make the Poolsheen program an executable file.

By making the Poolshen program a JAR file, it can be executed by
any other users who have a JDK.

lfrostbytee committed:
Add JavaDocs for all files

lfrostbytee committed:
Add Coding standard to all files

lfrostbytee committed:
Add Find Command

lfrostbytee committed:
Merge branch 'branch-A-JavaDoc'

lfrostbytee committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/poolsheen/Parser.java
#	src/main/java/poolsheen/Ui.java

lfrostbytee committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/poolsheen/Parser.java

tinenhao commented on src/main/java/poolsheen/Storage.java:
```diff
@@ -0,0 +1,127 @@
+import poolsheen.task.Task;
+import poolsheen.task.Event;
+import poolsheen.task.Deadline;
+import poolsheen.task.ToDo;
```
Would this be better if ordered alphebetically?

lfrostbytee replied:
Thank you for pointing this out to me!

tinenhao commented on src/main/java/poolsheen/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+        return arl.isEmpty();
+    }
+
+    public int size() {
```
Should this name be a verb?

lfrostbytee replied:
Yup, definitely, since it is a method.

tinenhao commented on src/main/java/poolsheen/Poolsheen.java:
```diff
@@ -0,0 +1,92 @@
+                ui.showError("Poolsheen has never seen this command and is confused.",
+                        "IllegalArgumentException");
+            }  catch (Exception e) {
+                System.out.println("An unexpected error has occurred and the program will end.");
```
I like that you have differentiated the different exceptions to catch and print out the corresponding error message

tinenhao commented on src/main/java/poolsheen/Parser.java:
```diff
@@ -0,0 +1,59 @@
+            return new EventCommand(arl);
+        case "FIND":
+            return new FindCommand(arl);
+        default:
```
I like that you abstracted out each command which makes the code look much cleaner

ryanlml commented on src/main/java/poolsheen/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,41 @@
+            throw new IncompleteCommandException(String.join(" ", rest),
+                    "deadline", "deadline commands need a '/by'");
+        } else {
+            String descD = String.join(" ", rest.subList(0, rest.indexOf("/by")));
+            String timeD = String.join(" ", rest.subList(rest.indexOf("/by") + 1, rest.size()));
```
These two lines are a little tough for me to understand at first glance. I understand that writing everything in one compact line looks aesthetically nice, but it compromises on readability. May I suggest defining variables to for `rest.indexOf("/by")` and `rest.subList(...)`. Just a suggestion!

lfrostbytee replied:
Looks like a good suggestion to me, thank you!

ryanlml commented on src/main/java/poolsheen/command/FindCommand.java:
```diff
@@ -0,0 +1,28 @@
+            throw new IncompleteCommandException(String.join(" ", rest),
+                    "find", "The find command only accepts one word");
+        } else {
+            ui.displayList(tl.find(rest.get(0)));
```
Same issue here! The nested brackets makes it quite hard to understand what the code is supposed to do. Perhaps consider giving variable names to `rest.get(0)`, etc!

lfrostbytee replied:
Nice spot. I'll work on that.

ryanlml commented on src/main/java/poolsheen/Ui.java:
```diff
@@ -0,0 +1,143 @@
+ * Represents everything that involves interactions with the user.
+ */
+public class Ui {
+    /** The indentation for all printed Poolsheen lines. */
+    public static final String BEGIN_SPACE = "      ";
+
+    /** The Poolsheen logo. */
+    private static final String LOGO = "\n" + "POOLSHEEN";
+
+    /** The message first printed when the Poolsheen program runs. */
+    private static final String WELCOME_MESSAGE = "Hello from" + LOGO + "\n" +
+            "Type something in for Poolsheen to respond to you:";
+
+    /** The message printed last whenever Poolsheen replys. */
+    private static final String LAST_REPLY = "meow *_*";
+
+    /** A horiontal line to distinguish each interaction the user has with Poolsheen. */
+    private static final String HORIZONTAL_LINE = "---------------";
+
+    /** The message last printed when the Poolsheen program ends. */
+    private static final String GOODBYE_MESSAGE = "MeoAww... See you next time :(\n"
```
Great idea to abstract all string constants here! Makes the job easier should you want to change the UI messages at any point in the future. I also noted that you've capitalised the variable names; it's a common mistake that you've avoided. Good job!

ryanlml commented on src/main/java/poolsheen/Storage.java:
```diff
@@ -0,0 +1,127 @@
+     * @throws IOException Thrown if there is an error with parsing the file.
+     */
+    public ArrayList<Task> load() throws PoolsheenException {
+        ArrayList<Task> res = new ArrayList<>(100);
```
Perhaps the variable `res` could use a more precise name, like `tasks` or `listOfTasks`. Just a thought!

ryanlml commented on src/main/java/poolsheen/Ui.java:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @param errMsg The string obtained from an error's getMessage function.
+     */
+    public void showError(String errMsg, String errorType) {
+        System.out.println(HORIZONTAL_LINE + "\n" +
+                newLine("The following UI error has occurred") +
+                newLine("The error type is: " + errorType) +
+                newLine(errMsg) +
+                newLine("Please try again!") +
+                newLine(LAST_REPLY) + HORIZONTAL_LINE);
+    }
```
The abstraction of constant strings pays off well here! With the `newLine` function the code looks clean and readable, contrast to having a slew of `System.out.println` messages. Well done!

ryanlml reviewed (commented):
Your code is overall well done and very readable! I could not find any major issues with code quality and had to spend some time nitpicking on the details, but they are very minor issues. Good job!

lfrostbytee committed:
Correct mistakes pointed out by reviewers

lfrostbytee committed:
Add checkstyle

Currently, there are not many tools that can detect coding style
violations.

This needs to change as coding style standard and quality is
important for maintaining the project.

Let's add the Checkstyle tool.

The Checkstyle tool is a static analysis tool that can detect
errors in code before compilation. It prevents errors early.

Refer to this link for how the Checkstyle tool is implemented
https://se-education.org/guides/tutorials/checkstyle.html#configuring-checkstyle

lfrostbytee committed:
Add GUI

Currently, the Poolsheen program is a Command Line Interface
program.

We want to make the Poolsheen program more user friendly.

Let's change the Poolsheen program from a CLI to a GUI using
javaFX.

JavaFX offers simple tools to construct a simple GUI for the
Poolsheen program.

lfrostbytee committed:
Fix launcher

Previously, the poolsheen.jar file could not be executed because
the mainClassName was specified wrongly.

We want to fix the jar executable so that the Poolsheen program
works.

Let's change the path specified in mainClassName to
poolsheen.Launcher.

This ensures that all executables will always run from the Launcher
class first.

lfrostbytee committed:
Add Java Assertions

Currently, only JUnit Assertions are implemented in the code. If an
error occurs, it has to be caught by any try-catch blocks during
run time.

Let's implement Java assertions.

Java assertions can ensure that assumptions are truly met during
run-time.

lfrostbytee committed:
Fix code quality

Currently the code quality is not adequate even after making fixes
suggested by other classmates.

We want to improve the code quality following the suggestions given
in the Week 5 Topics.

Let's follow the Code Quality guidelines provided by Week 5.

Guard clauses have been added to improve code readability.
New boolean names have been added to make conditionals easier to
understand.
Arrowhead code have been reduced.
simpli

lfrostbytee committed:
Add Streams

Currently, most operations rely on Array List operations.

However, such operations may be hard to read. And may take more
time to re-write, so new operations are needed.

Let's rewrite the Find method using Streams.

The Stream functions, Predicate and Function classes are used to
write more intuitive  code.

lfrostbytee committed:
Merge pull request #1 from lfrostbytee/branch-A-Streams

Add Streams

lfrostbytee committed:
Merge branch 'master' into branch-A-CodeQuality

lfrostbytee committed:
Merge branch 'master' into branch-A-Assertions

lfrostbytee committed:
Merge pull request #2 from lfrostbytee/branch-A-Assertions

Branch a assertions

lfrostbytee committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/poolsheen/Storage.java

lfrostbytee committed:
Merge pull request #3 from lfrostbytee/branch-A-CodeQuality

Branch a code quality

lfrostbytee committed:
Add Update command

Currently, there is no way to update tasks without adding or
deleting them.

Users should have a more convenient way to change the task time or
description.

Let's create an Update command.

We can now update tasks by calling
"update [pos] [desc/time] [content]" .
For example "update 1 desc newDesc" updates the task at position 1
of the task list with a new description called newDesc.

lfrostbytee committed:
Merge pull request #4 from lfrostbytee/branch-C-Update

Add Update command

lfrostbytee committed:
Add Continuous Integrations

lfrostbytee committed:
Remove IO Redirection Tests

Currently several GithubActions fail during the merge pull request.

This could be due to the IO Redirection tests which are not
applicable for the Poolsheen program in its GUI State.

Let's remove the IO Redirection Tests by editting the gradle.yml
file.

lfrostbytee committed:
Merge pull request #6 from lfrostbytee/branch-A-CI

Branch a ci

lfrostbytee committed:
Improve GUI

Currently, the GUI for the Poolsheen program is stale and not
aesthetically pleasing.

Users may be turned off given the unaesthetic nature of the
Poolsheen program.

Let's add more CSS styles into the Dialog Box and Main Window.

lfrostbytee committed:
Merge pull request #8 from lfrostbytee/branch-A-BetterGui

Improve GUI

lfrostbytee committed:
Improve Poolsheen personality

Currently, the replies from Poolsheen are not interesting.

Users may stop using the Poolsheen program as they find it boring.

Let's improve the replies of Poolsheen.

Poolsheen now has a warning line when errors are found. Poolsheen
replies now have a horizontal line that fits the window. Poolsheen
now says certain cat messages.

lfrostbytee committed:
Reduce line length

One of the code lines in Storage.java is too long and failed a
Checkstyle Test.

lfrostbytee committed:
Merge pull request #9 from lfrostbytee/branch-A-Personality

Branch a personality

lfrostbytee committed:
Improve Error Handling

Currently, some error cases cannot be handled.

These errors cause the program to crash and stop running.

Let's increase the number of error handling cases.

For instance, let's handle the case where the date and time parsed
is incorrect even if it is in the correct format.

lfrostbytee committed:
Merge pull request #10 from lfrostbytee/branch-A-MoreErrorHandling

Improve Error Handling

lfrostbytee committed:
Add more JUnit Tests

Some new functions do not have JUnit Tests yet.

To improve maintainability, we need to introduce more JUnit Tests.

Let's add new JUnit Tests for Ui and Parser.

lfrostbytee committed:
Merge pull request #11 from lfrostbytee/branch-A-MoreTesting

Add more JUnit Tests

lfrostbytee committed:
Improve Code Quality

lfrostbytee committed:
Add User Guide and product screenshot

Currently, the user guide is not updated.

Users may not be interested in using the Poolsheen program as
they are not sure on how to use it.

Let's make a User Guide using Github Flavored Markdown.

lfrostbytee committed:
Merge pull request #12 from lfrostbytee/branch-A-UserGuide

Add User Guide and product screenshot

lfrostbytee committed:
Update version number

lfrostbytee committed:
Improve code quality and improve error handling

Previously there was dead code located in Poolsheen.java. Also,
the program was unable to handle the scenario where non-positive
numbers were inputted into commands.

Users might not be interested in using the Poolsheen program if it
does not work.

Let's remove dead code and add more error handling cases.

lfrostbytee committed:
Update code quality and README.md

Added new comments to thank authors who inspired certain codes.
Also improved README.md with emojis.

lfrostbytee committed:
Fix README.md and fonts for Mac Users

lfrostbytee committed:
Fix fonts for Mac Users
